### PR TITLE
Task bot 1271 worker export

### PR DESCRIPTION
**Pull request by tchevily**
# Description

This PR adds impacts exporter class

## Issue (optional)

Issue link: BOT-1271

## Pull Request type (optional)

This is a new feature

## How to test

Here are the following steps to test this pull request:

- In terminal from root path launch ```python impactsExporter.py -c xxxx-yyyy-zzzz```
Where xxxx-yyyy-zzzz is a valid user Uuid

### CODE REVIEW

File: impactsExporter.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    def createCSV(self, filePath, impacts):
+        with open(filePath, 'wb') as f:
+            outcsv = csv.writer(f)
+            outcsv.writerow(impacts.keys())
+            outcsv.writerows(impacts.fetchall())
+            f.close()
+
+    def markExportAsDone(self, item, filePath):
+        item.status = 'done'
+        item.file_path = filePath
```
Comment (Fahmus) [inline impactsExporter.py]: item.updated_at = utils.get_current_time()

File: impactsExporter.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    def get_oldest_waiting_export(self, clientId):
+        item = Export.get_oldest_waiting_export(clientId)
+        if not item:
+            raise ValueError("No waiting task for this client has been found")
+
+        return item
+
+    def update_export_status(self, item, status):
+        item.status = status
+        item.process_start_date = utils.get_current_time()
```
Comment (Fahmus) [inline impactsExporter.py]: item.updated_at = utils.get_current_time()

File: impactsExporter.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+class impactsExporter:
+
+    def setclientId(self, id):
+        self.clientId = id
+        self.logger = logging.getLogger('impacts exporter')
+
+    def get_oldest_waiting_export(self, clientId):
+        item = Export.get_oldest_waiting_export(clientId)
+        if not item:
+            raise ValueError("No waiting task for this client has been found")
```
Comment (Fahmus) [inline impactsExporter.py]: why you raise Error? if there is no waiting task continue to the next client or leave

File: impactsExporter.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    def run(self):
+        try:
+            self.checkRequiredArguments()
+            item = self.get_oldest_waiting_export(self.client.id)
+            self.update_export_status(item, 'handling')
+            impacts = self.getArchivedImpacts(item.client_id, item.start_date, item.end_date)
+            filePath = self.generateFilePath(item)
+            self.createCSV(filePath, impacts)
+            self.markExportAsDone(item, filePath)
+            self.logger.info('Impacts are exported in ' + filePath)
```
Comment (dvdn) [inline impactsExporter.py]:
```suggestion
            self.logger.info('Impacts export for {} is available at {}'.format(item.client_id, filePath))
```

File: chaos/utils.py [hunk 5]
```diff
@@ -871,3 +871,8 @@ def add_notification_date_on_impacts(disruptions_or_impacts_json):
         if ('send_notifications' in impact) and impact['send_notifications'] and \
             ('notification_date' not in impact):
             impact['notification_date'] = get_current_time().strftime('%Y-%m-%dT%H:%M:%SZ')
+
+def is_uuid(uuid_string):
+    import re
```
Comment (dvdn) [inline chaos/utils.py]: You can move it at the beginning of the file

File: chaos/models.py [hunk 6]
```diff
@@ -2055,3 +2017,89 @@ def get(cls, client_id, id=None, key=None, type=None):
...
+                ' disruption d' \
+                ' LEFT JOIN impact i ON (i.disruption_id = d.id)' \
+                ' LEFT JOIN application_periods app ON (app.impact_id = i.id)' \
+                ' LEFT JOIN cause c ON (c.id = d.cause_id)' \
+                ' LEFT JOIN associate_impact_pt_object aipto ON (i.id = aipto.impact_id)' \
+                ' LEFT JOIN pt_object po ON (aipto.pt_object_id = po.id)' \
+                ' WHERE' \
+                ' d.client_id = :client_id' \
+                ' AND app.start_date >= :app_start_date' \
+                ' AND app.end_date <= :app_end_date'
```
Comment (pthegner) [inline chaos/models.py]: I know that the function is named `getArchivedImpacts` but i don't think that you have to also get the impacts with `archived`  status. Maybe you should had a filter here.

File: chaos/resources.py [hunk 7]
```diff
@@ -1894,7 +1895,7 @@ class ImpactsExports(flask_restful.Resource):
     def __init__(self):
         self.parsers = {'get': reqparse.RequestParser()}
 
-    def _validate_dates_boundary(self, json):
+    def validate_dates_boundary(self, json):
```
Comment (Netmisa) [inline chaos/resources.py]: Some times you use camelCase for function in this class, and some time snake_case

File: chaos/utils.py (hunk 5 above)
Comment (tchevily) [reply chaos/utils.py]: why ?

File: chaos/resources.py [hunk 8]
```diff
@@ -1903,6 +1904,20 @@ def _validate_dates_boundary(self, json):
...
+        from subprocess import Popen, PIPE
+        from os import path
+        from sys import executable
+
+        root_dir = path.abspath(path.join(__file__, "../.."))
+        exporter_path = path.join(root_dir, 'impactsExporter.py')
+
+        p1 = Popen([executable, exporter_path, '-c', client_id], stdout=PIPE)
+
+        logging.getLogger(__name__).debug(p1.communicate()[0])
```
Comment (pthegner) [inline chaos/resources.py]: It's maybe the communicate here that wait for the process end : https://stackoverflow.com/questions/52510318/send-input-to-python-subprocess-without-waiting-for-result

**Comment by pthegner** [comment]
There is just one problem with the timezone -> you should use it for generate the file name.

File: impactsExporter.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    parser.add_argument('--token', help='Navitia token')
+    parser.add_argument('--tz', help='Time zone')
+
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    args = get_command_arguments(sys.argv[1:])
+
+    tz = args.tz;
```
Comment (Fahmus) [inline impactsExporter.py]: Unnecessary semicolon
```suggestion
    tz = args.tz
```

File: chaos/utils.py (hunk 5 above)
Comment (Fahmus) [reply chaos/utils.py]: why?

Comment (Fahmus) [reply chaos/utils.py]: All imports are at the beginning of the file :)

File: chaos/resources.py [hunk 10]
```diff
@@ -1938,5 +1956,33 @@ def post(self, client):
...
+            self.run_export(export)
 
         return marshal({'export': export}, one_export_fields), 201
+
+
+class ImpactsExportDownload(flask_restful.Resource):
+    def __init__(self):
+        self.parsers = {'get': reqparse.RequestParser()}
+
+    @validate_client()
```
Comment (Fahmus) [inline chaos/resources.py]: @validate_client_token()

Comment (Fahmus) [inline chaos/resources.py]: @validate_client_token()

File: chaos/models.py [hunk 11]
```diff
@@ -2055,3 +2017,129 @@ def get(cls, client_id, id=None, key=None, type=None):
...
+                ' LEFT JOIN message m ON (i.id = m.impact_id and m.channel_id = ch.id) ' \
+                ' LEFT JOIN pattern ptrn ON (i.id = ptrn.impact_id) ' \
+                ' WHERE' \
+                ' d.client_id = :client_id' \
+                ' AND ch.client_id = :client_id'\
+                ' AND app.start_date >= :app_start_date' \
+                ' AND app.end_date <= :app_end_date' \
+                ' AND c.is_visible = :is_visible' \
+                ' AND ch.is_visible = :is_visible' \
+                ' GROUP BY ' \
```
Comment (pthegner) [inline chaos/models.py]: The GROUP BY is not necessary

File: chaos/models.py [hunk 12]
```diff
@@ -2055,3 +2017,129 @@ def get(cls, client_id, id=None, key=None, type=None):
...
+                ', po.uri AS pt_object_name' \
+                ', s.wording AS severity' \
+                ', (CASE WHEN cht.name=\'title\' THEN m.text ELSE \'\' END) as impact_title' \
+                ', i.status' \
+                ', ch.name AS channel_name' \
+                ', cht.name AS channel_type_name' \
+                ', m.text AS channel_message' \
+                ', app.start_date AS application_start_date' \
+                ', app.end_date AS application_end_date' \
+                ', (CASE WHEN (SELECT COUNT(1) FROM application_periods app_period WHERE app_period.impact_id = i.id AND app_period.start_date >= :app_start_date AND app_period.end_date <= :app_end_date) > 1 THEN True ELSE False END) AS periodicity' \
```
Comment (pthegner) [inline chaos/models.py]: app_start_date and app_end_date is not necessary here (we want to know if the impact is periodic, not just between the date)

File: impactsExporter.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+        columns = impacts.keys()
+
+        rows = []
+        for sub_dict in impacts.fetchall():
+            row = []
+            for column in columns:
+                val = sub_dict[column]
+                if column == 'pt_object_name' :
+                    val = navitia.find_tc_object_name(sub_dict['pt_object_uri'], sub_dict['pt_object_type'])
+                elif column == 'periodicity':
```
Comment (pthegner) [inline impactsExporter.py]: you should maybe add this : 
```
                elif column == 'channel_message' and val is not None:
                    val = val.replace('\n', ' ').replace('\r', '')
```
in order to delete the line breaks for certain messages (view with Victor)

**Pull request merged**